rustls/src/client/client_conn.rs

```diff
@@ -386,6 +389,93 @@ pub enum Tls12Resumption {
     SessionIdOrTickets,
 }
 
+/// Configuration for performing encrypted client hello.
+///
+/// Note: differs from the protocol-encoded EchConfig (`EchConfigMsg`).
+#[derive(Clone, Debug)]
+#[allow(dead_code)] // TODO(@cpu): remove in subsequent commit.
+pub struct EchConfig {
+    /// The provider to use for HPKE operations.
+    pub(crate) hpke_provider: &'static dyn HpkeProvider,
+
+    /// The selected EchConfig.
+    pub(crate) config: EchConfigMsg,
+
+    /// An HPKE suite from the `config` we have selected as compatible with the `hpke_provider`.
+    pub(crate) suite: HpkeSuite,
+}
+
+impl EchConfig {
+    /// Construct an EchConfig by selecting a ECH config from the provided bytes that is compatible
+    /// with the given HPKE provider.
+    ///
+    /// The config bytes should be sourced from a DNS-over-HTTPS lookup resolving the `HTTPS` service
+    /// record record type for the host name of the server you wish to connect to using ECH,
+    /// and extracting the ECH service record parameter.
+    ///
+    /// One of the provided configurations must be compatible with the HPKE provider's supported
+    /// suites or an error will be returned.
+    // TODO(@cpu): update docstring with a pointer to the example program using hickory-dns.
+    // TODO(@cpu): consider dedicated HpkeError type(s).
+    pub fn new(
+        ech_configs: impl AsRef<[u8]>,
```

Comment:
Hmm, I don't love the very generic parameter type here. Here's where it might be interesting to have rustls-pki-types type that represents a `Vec<EchConfigMsg>` somehow such that hickory could yield that type? (Maybe too ambitious for this stage.)